cache/cache.cpp

```diff
@@ -190,11 +194,332 @@ size_t cache_manager::idx(const unsigned char *id) {
+                                  struct dnet_net_state *st,
+                                  struct dnet_cmd *cmd,
+                                  struct dnet_io_attr *io)
+{
+	struct dnet_node *n = st->n;
+
+	auto it = cache->read(io->id, io->flags);
+	if (!it) {
+		if (!(io->flags & DNET_IO_FLAGS_CACHE)) {
+			return -ENOTSUP;
+		}
+
+		return -ENOENT;
+	}
+
+	auto d = it->data;
+
+	/*!
+	 * When offset is larger then size of the file, operation is definitely incorrect
+	 */
+	if (io->offset >= d->size()) {
+		BH_LOG(*n->log, DNET_LOG_ERROR, "%s: %s cache: invalid offset: "
+		       "offset: %llu, size: %llu, cached-size: %zd",
+		       dnet_dump_id(&cmd->id), dnet_cmd_string(cmd->cmd),
+		       (unsigned long long)io->offset, (unsigned long long)io->size,
+		       d->size());
+		return -EINVAL;
+	}
+
+	/*!
+	 * If offset is correct, but offset + read_size is bigger then file_size
+	 * then we should return data from offset position till the end of the file
+	 * This situation happens when for example we want to read first 100 bytes of
+	 * the file and it's size appears to be less then 100 bytes.
+	 */
+	io->size = std::min(io->size, d->size() - io->offset);
+
+	/*!
+	 * 0 is special value for io operation size and in this case we should read all file
+	 */
+	if (io->size == 0)
+		io->size = d->size() - io->offset;
+
+	io->total_size = d->size();
+
+	io->timestamp = it->timestamp;
+	io->user_flags = it->user_flags;
+
+	cmd->flags &= ~DNET_FLAGS_NEED_ACK;
+	return dnet_send_read_data(st, cmd, io, &d->at(io->offset), -1, io->offset, 0);
+}
+
+static int dnet_cmd_cache_io_read_new(struct cache_manager *cache,
+                                      struct dnet_net_state *st,
+                                      struct dnet_cmd *cmd,
+                                      void *data)
+{
+	using namespace ioremap::elliptics;
+
+	auto request = [&data, &cmd] () {
+		dnet_read_request request;
+		deserialize(data_pointer::from_raw(data, cmd->size), request);
+		return request;
+	} ();
+
+	if (request.ioflags & DNET_IO_FLAGS_NOCACHE) {
+		return -ENOTSUP;
+	}
+
+	auto it = cache->read(reinterpret_cast<unsigned char *>(&cmd->id.id), request.ioflags);
+	if (!it) {
+		if (!(request.ioflags & DNET_IO_FLAGS_CACHE)) {
+			return -ENOTSUP;
+		}
+
+		return -ENOENT;
+	}
+
+	auto raw_data = it->data;
+	auto raw_json = it->json;
+
+	data_pointer json, data_p;
+
+	if (request.read_flags & DNET_READ_FLAGS_JSON) {
+		json = data_pointer::from_raw(*raw_json);
+	}
+
+	uint64_t data_size = 0;
+
+	if (request.read_flags & DNET_READ_FLAGS_DATA) {
+		data_size = raw_data->size();
+
+		if (request.data_offset >= data_size)
+			return -E2BIG;
+
+		if (request.data_size &&
+		    request.data_size < data_size)
+			data_size = request.data_size;
+
+		data_p = data_pointer::from_raw(*raw_data);
+		data_p = data_p.skip(request.data_offset);
+	}
+
+	auto header = serialize(dnet_read_response{
+		0,
+		it->user_flags,
+
+		it->json_timestamp,
+		raw_json->size(),
```

Comment:
1. raw_json can't be nullptr since it is initialized in data_t ctor.
2. json may be uninitialized (when DNET_READ_FLAGS_JSON flag is not set), so json.size() will return 0, while raw_json is not empty.